File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque

def solution(maps):
    def bfs(start_x, start_y, end_x, end_y):
        queue = deque()
        queue.append([start_x, start_y, 0])
        visited = set()
        visited.add((start_x, start_y))

        while queue:
            cur_x, cur_y, count = queue.popleft()

            if cur_x == end_x and cur_y == end_y:
                return count
            
            for i in range(4):
                cur_dx = cur_x + dx[i]
                cur_dy = cur_y + dy[i]

                if 0 <= cur_dx < n and 0 <= cur_dy < m and maps[cur_dx][cur_dy] != 'X':
                    if (cur_dx, cur_dy) not in visited:
                        visited.add((cur_dx, cur_dy))
                        queue.append([cur_dx, cur_dy, count + 1])

    n = len(maps)
    m = len(maps[0])
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    for i in range(n):
            for j in range(m):
                if maps[i][j] == 'S':
                    x, y = i, j
                elif maps[i][j] == 'E':
                    end_x, end_y = i, j
                elif maps[i][j] == 'L':
                    l_x, l_y = i, j
                    
    toLever = bfs(x, y, l_x, l_y)
    toEnd = bfs(l_x, l_y, end_x, end_y)
    
    if toLever and toEnd:
        return toLever + toEnd
    
    return -1

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    private static String[] maze;
    private static boolean[][] visited;
    private static int row, col;
    private static ArrayDeque<int[]> queue;
    private static int[] dx = {-1, 1, 0, 0};
    private static int[] dy = {0, 0, -1, 1};
    
    private static int[] findDestination(char ch) {
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (maze[i].charAt(j) == ch) return new int[] {i, j};
            }
        }
        return new int[] {0, 0};
    }
    
    private static int toDestination(int[] from, int[] to) {
        visited = new boolean[row][col];
        queue = new ArrayDeque<>();
        
        int startX = from[0];
        int startY = from[1];
        int toX = to[0];
        int toY = to[1];
        
        queue.addLast(new int[] {startX, startY, 0});
        visited[startX][startY] = true;
        
        while (!queue.isEmpty()) {
            int[] arr = queue.pollFirst();
            int curX = arr[0];
            int curY = arr[1];
            int count = arr[2];
            
            if (curX == toX && curY == toY) return count;
            
            for (int i = 0; i < 4; i++) {
                int nextX = curX + dx[i];
                int nextY = curY + dy[i];
                
                if (nextX >= 0 && nextX < row && nextY >= 0 && nextY < col) {
                    if (!visited[nextX][nextY] && maze[nextX].charAt(nextY) != 'X') {
                        visited[nextX][nextY] = true;
                        queue.addLast(new int[] {nextX, nextY, count + 1});
                    }
                }
            }
        }
        
        return -1;
    }
    
    
    public int solution(String[] maps) {
        /*
        maps 최대 100 X 100
        
        [풀이]
        레버를 당기고 탈출 가능
        레버까지 최단 거리 구하고, 레버에서 탈출구까지 최단 거리 구해서 합하기
        */
        maze = maps;
        row = maps.length;
        col = maps[0].length();

        
        int[] start = findDestination('S');
        int[] lever = findDestination('L');
        int[] exit = findDestination('E');
        
        int toLever = toDestination(start, lever);
        int toExit = toDestination(lever, exit);
        
        if (toLever == -1 || toExit == -1) return -1;
        
        return toLever + toExit;
    }
}